File: daily_problems/2025/06/0619/personal_submission/cf575b_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n;
    cin >> n;

    vector<int> us, vs, ts;
    vector<vector<int>> path(n);

    for (int eid = 0; eid < n - 1; eid ++) {
        int u, v, t;
        cin >> u >> v >> t;

        u --, v --;
        us.emplace_back(u);
        vs.emplace_back(v);
        ts.emplace_back(t);

        path[u].emplace_back(eid);
        path[v].emplace_back(eid);
    }


    vector<vector<int>> nth_parent(20, vector<int>(n, -1));
    vector<int> parent_eid(n), depth(n, 0), order;

    auto dfs = [&] (auto &self, int u, int p) -> void {
        order.emplace_back(u);

        for (auto &eid: path[u]) {
            int v = us[eid] + vs[eid] - u;
            if (v != p) {
                nth_parent[0][v] = u;
                parent_eid[v] = eid;
                depth[v] = depth[u] + 1;
                self(self, v, u);
            }
        }
    };

    dfs(dfs, 0, -1);

    for (int i = 1; i < 20; i ++) {
        for (int u = 0; u < n; u ++) {
            if (nth_parent[i - 1][u] != -1) {
                nth_parent[i][u] = nth_parent[i - 1][nth_parent[i - 1][u]];
            }
        }
    }

    auto lca = [&] (int u, int v) -> int {
        if (depth[u] < depth[v]) swap(u, v);

        int d = depth[u] - depth[v];

        for (int i = 0; i < 20; i ++) {
            if (d >> i & 1) {
                u = nth_parent[i][u];
            }
        }

        if (u == v) return u;

        for (int i = 19; i >= 0; i --) {
            if (nth_parent[i][u] != nth_parent[i][v]) {
                u = nth_parent[i][u];
                v = nth_parent[i][v];
            }
        }

        return nth_parent[0][u];
    };

    vector<int> ups(n, 0), downs(n, 0);

    int mod = 1e9 + 7;

    auto add = [&] (int &x, int y) -> void {
        x += y;
        if (x >= mod) x -= mod;
    };

    int k;
    cin >> k;

    vector<int> pw2(k + 1, 0); pw2[0] = 1;
    for (int i = 1; i <= k; i ++) {
        pw2[i] = pw2[i - 1];
        add(pw2[i], pw2[i - 1]);
    }

    int u = 0;
    while (k --) {
        int v;
        cin >> v;
        v --;

        int l = lca(u, v);
        ups[u] ++;
        ups[l] --;
        downs[v] ++;
        downs[l] --;

        u = v;
    }

    reverse(order.begin(), order.end());
    for (auto &i: order) {
        if (i) {
            ups[nth_parent[0][i]] += ups[i];
            downs[nth_parent[0][i]] += downs[i];
        }
    }

    int ans = 0;
    for (int i = 1; i < n; i ++) {
        if (ts[parent_eid[i]]) {
            add(ans, pw2[(i == us[parent_eid[i]] ? downs[i] : ups[i])] - 1);
        }
    }

    cout << ans;

    return 0;
}

File: daily_problems/2025/06/0619/personal_submission/cf575b_yawn_sean.py
# Submission link: https://codeforces.com/problemset/submission/575/325082678#
def main():
    n = II()
    us = []
    vs = []
    ts = []

    path = [[] for _ in range(n)]

    for eid in range(n - 1):
        u, v, t = MII()
        u -= 1
        v -= 1
        
        us.append(u)
        vs.append(v)
        ts.append(t)
        path[u].append(eid)
        path[v].append(eid)

    parent = [-1] * n
    parent_eid = [0] * n
    depth = [0] * n
    que = [0]

    for u in que:
        for eid in path[u]:
            v = us[eid] + vs[eid] - u
            if parent[u] != v:
                parent[v] = u
                parent_eid[v] = eid
                depth[v] = depth[u] + 1
                que.append(v)

    nth_parent = [[-1] * n for _ in range(20)]
    nth_parent[0] = parent

    for i in range(1, 20):
        for u in range(n):
            if nth_parent[i - 1][u] != -1:
                nth_parent[i][u] = nth_parent[i - 1][nth_parent[i - 1][u]]

    def lca(u, v):
        if depth[u] < depth[v]:
            u, v = v, u
        
        d = depth[u] - depth[v]
        while d:
            x = d & -d
            
            bit = x.bit_length() - 1
            u = nth_parent[bit][u]
            
            d -= x
        
        if u == v: return u
        
        for i in range(19, -1, -1):
            if nth_parent[i][u] != nth_parent[i][v]:
                u = nth_parent[i][u]
                v = nth_parent[i][v]
        
        return nth_parent[0][u]

    ups = [0] * n
    downs = [0] * n

    k = II()
    nodes = LGMI()

    u = 0
    for i in range(k):
        v = nodes[i]
        
        l = lca(u, v)
        
        ups[u] += 1
        ups[l] -= 1
        downs[v] += 1
        downs[l] -= 1
        u = v

    for i in reversed(que):
        if i > 0:
            ups[parent[i]] += ups[i]
            downs[parent[i]] += downs[i]

    mod = 10 ** 9 + 7
    pw2 = [1] * (k + 1)

    for i in range(1, k + 1):
        pw2[i] = pw2[i - 1] + pw2[i - 1]
        if pw2[i] >= mod: pw2[i] -= mod

    ans = 0
    for i in range(1, n):
        if ts[parent_eid[i]]:
            if i == us[parent_eid[i]]: ans += pw2[downs[i]] - 1
            else: ans += pw2[ups[i]] - 1
            if ans >= mod: ans -= mod

    print(ans)